cmd, pkg/podman: Introduce Image Interface and Unit Tests

DaliborKr opened the pull request:
Introduces an Image interface based on the conversation in PR #1707.

Similar to how container information is handled (introduced in commits e611969 and ec7eb59), this PR abstracts the representation of an image. This is necessary because `podman inspect --type image` and `podman images` return different JSON structures. The Image interface provides a unified way to access image data, with two distinct implementations to handle the different JSON formats.


Updates related functions:
- GetImages() now returns an Images iterator, which handles image flattening and sorting internally.
- InspectImage() now returns an Image interface instead of a generic map[string]interface{}.
- The IsToolboxImage() function has been replaced with an IsToolbx() method on the Image interface.


Adds unit tests: This PR includes unit tests for the new Image interface implementation, covering the output of both `podman inspect --type image` and `podman images`.

debarshiray commented on src/pkg/podman/image.go:
```diff
@@ -0,0 +1,258 @@
+}
+
+// ALREADY EXISTS IN CONTAINER.GO
+//
```
I guess we need to take a decision about the code here in comments?  :)

debarshiray reviewed (changes requested):
Thanks for working on this, @DaliborKr !  My apologies for the delay.

I am still reading through the changes, but one thing popped out at me.

debarshiray reviewed (changes requested):
I noticed a few more relatively minor things while trying to rebase this against `main`.

DaliborKr commented on src/pkg/podman/image.go:
```diff
@@ -0,0 +1,258 @@
+	ID() string
+	IsToolbx() bool
+	Labels() map[string]string
+	Names() []string
```
This is a very good idea! We can have a method `Name()` like the example below, which also encapsulates the assertion, and use the `Names()` method only to flatten the names within the `podman` package, ensuring that only flattened images are returned from its functions. 

```
func (image *imageImages) Name() string {
    if len(image.names) != 1 {
        panic("cannot use Name() on unflattened Image")
    }
    return image.names[0]
}
```


Did you mean something like this?

DaliborKr committed:
cmd/rmi, pkg/podman: Make podman.InspectImage() return an Image

Unmarshal the JSON from 'podman inspect --format json --type image'
directly inside podman.InspectImage() to confine the details within the
podman package.

The JSON samples for the unit tests were taken using the default Toolbx
image on versions of Fedora that shipped a specific Podman and Toolbx
version.  This accounts for differences in the JSON caused by different
major versions of Podman and the way different Toolbx images were built.

One exception was Fedora 28, which had Podman 1.1.2 and Toolbx 0.0.9,
which was the last Toolbx version before 'toolbox init-container' became
the entry point for all Toolbx containers [1].  However, the default
Toolbx image is no longer available from registry.fedoraproject.org.
Hence, the image for Fedora 29 was used.

The minimum required Podman version is 1.6.4 [2], and the Go
implementation has been encouraging users to create containers with
Toolbx version 0.0.97 or newer [3].  The versions used to collect the
JSON samples for the unit tests were chosen accordingly.  They don't
exhaustively cover all possible supported and unsupported version
combinations, but hopefully enough to be useful.

The fedora-toolbox image went through a series of significant changes
during the Fedora 39 and 40 development cycles.  Fedora 38 was the last
release where the image was built from a Container/Dockerfile [4].  For
Fedora 39, it was rewritten in terms of fedora-kickstarts and
pungi-fedora for the ToolbxReleaseBlocker Change [5].  For Fedora 40, it
was rewritten as a KIWI description as part of the KiwiBuiltCloudImages
Change [6].  Hence, all three Fedora versions were chosen to cover this
transition.

[1] Commit 8b84b5e4604921fa
    https://github.com/containers/toolbox/commit/8b84b5e4604921fa
    https://github.com/containers/toolbox/pull/160

[2] Commit 8e80dd5db1e6f40b
    https://github.com/containers/toolbox/commit/8e80dd5db1e6f40b
    https://github.com/containers/toolbox/pull/1253

[3] Commit af1216b2720c7ab5
    https://github.com/containers/toolbox/commit/af1216b2720c7ab5
    https://github.com/containers/toolbox/pull/1697
    https://github.com/containers/toolbox/issues/1684

[4] https://src.fedoraproject.org/container/fedora-toolbox

[5] https://fedoraproject.org/wiki/Changes/ToolbxReleaseBlocker

[6] https://fedoraproject.org/wiki/Changes/KiwiBuiltCloudImages

https://github.com/containers/toolbox/pull/1724

Signed-off-by: Dalibor Kricka <dalidalk@seznam.cz>

debarshiray commented on src/pkg/podman/image.go:
```diff
@@ -0,0 +1,261 @@
+
+	// Until Podman 2.0.x the field 'Created' held a human-readable string in
+	// format "5 minutes ago". Since Podman 2.1 the field holds an integer with
+	// Unix time. Go interprets numbers in JSON as float64.
```
On the surface, this comment about the field `Created` holding a human-readable string like *5 minutes ago* until Podman 2.0.x doesn't match the reality exposed by the tests.  :)

Speaking from memory, I think there are three cases.  Older than Podman 2.0, the Podman 2.0.x releases, and Podman 2.1 onwards.  We have to check this because I am not entirely sure.

Anyway, it's not the fault of this PR.  So there's no need to block on fixing this.  It's a note to ourselves to do something about this dissonance.  :)

debarshiray commented on src/pkg/podman/imageImages_test.go:
```diff
@@ -0,0 +1,792 @@
+				"			\"registry.fedoraproject.org/f29/fedora-toolbox:29\"" +
+				"		]," +
+				"		\"digest\": \"sha256:f324a546a5e894af041eea47f8f2392bf2a9e2d88ee77199b25a129174b7a0e1\"," +
+				"		\"created\": \"2019-11-06T11:36:10.586442Z\"," +
```
The `created` field in Podman older than 2.0.x holds an [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) formatted string, not something like *5 minutes ago*.

debarshiray commented on src/pkg/podman/imageImages_test.go:
```diff
@@ -0,0 +1,792 @@
+				"			\"sha256:1d1077ce12b6d45990709ab81543884838d71a3b2014d023b82fd465a3ce4e23\"" +
+				"		]," +
+				"		\"Created\": 1619352159," +
+				"		\"CreatedAt\": \"2021-04-25T12:02:39Z\"" +
```
Then, `Created` started holding a UNIX timestamp, and the [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) formatted string moved to `CreatedAt`.

debarshiray commented on src/pkg/podman/imageImages_test.go:
```diff
@@ -0,0 +1,792 @@
+						"maintainer":                     "Debarshi Ray \u003crishi@fedoraproject.org\u003e",
+						"name":                           "fedora-toolbox",
+						"release":                        "14",
+						"summary":                        "Base image for creating Fedora toolbox containers",
```
Nitpick: we can omit some of these unimportant labels that make the lines too wide.  :)

debarshiray commented on src/pkg/podman/imageImages_test.go:
```diff
@@ -0,0 +1,792 @@
+			data: "" +
+				"[" +
+				"	{" +
+				"		\"Id\": \"862705390e8b1678bbac66beb30547e0ef59abd65b18e23ea533f059ba069227\"," +
```
Nitpick: we can indent the JSON with 2 spaces instead of a tab to prevent the lines from becoming too wide.

debarshiray commented on src/pkg/podman/imageImages_test.go:
```diff
@@ -0,0 +1,792 @@
+				"		\"RepoTags\": null," +
+				"		\"RepoDigests\": [" +
+				"			\"sha256:b5f419099423fae80421cda692b00e301894101575806f44558e6c9c911769e6\"," +
+				"			\"sha256:bdab10512284f81480235e89939babff43b2a8ca5af870dbebdeb127c6533568\"" +
```
Nitpick: we can drop the `RepoDigests` because we don't use them anywhere and they make the lines too wide.

debarshiray commented on src/pkg/podman/imageImages_test.go:
```diff
@@ -0,0 +1,792 @@
+			imageCnt: 1,
+		},
+		{
+			name: "podman 5.6.1, fedora-toolbox:41",
```
Covering Fedora 40, 41 and 42 seems a bit wasteful because the Podman versions are likely to be similar.  Or did you have a specific reason for covering them?

There might be more value if we spaced them out.  eg., Fedora 40, 43, etc..  It's not a big deal.  We can just stay with Fedora 40 in this PR, like we have for the `containerInspect` tests, and extend the list later in future as more Fedora versions come out.

Strictly speaking, there can be some value in covering Fedora 38, 39 and 40, as three consecutive releases, because the `fedora-toolbox` images went through some major changes in each of them.  Fedora 38 was the last release where the image was built from a [Container/Dockerfile](https://src.fedoraproject.org/container/fedora-toolbox).  In Fedora 39, it moved to [fedora-kickstarts](https://pagure.io/fedora-kickstarts/) and [pungi-fedora](https://pagure.io/pungi-fedora/) as part of the [ToolbxReleaseBlocker](https://fedoraproject.org/wiki/Changes/ToolbxReleaseBlocker) Change.  In Fedora 40, it moved to [fedora-kiwi-descriptions](https://pagure.io/fedora-kiwi-descriptions) as part of the [KiwiBuiltCloudImages](https://fedoraproject.org/wiki/Changes/KiwiBuiltCloudImages) Change.

debarshiray replied:
I took the liberty to change the test cases using Fedora 41 and 42 to use Fedora 40, and added one with Fedora 39.

debarshiray commented on src/pkg/podman/imageImages_test.go:
```diff
@@ -0,0 +1,792 @@
+			imageCnt: 1,
+		},
+		{
+			name: "podman 5.6.1, fedora-toolbox:41, no name specified",
```
How was the JSON for this was generated?  Just curious.

I can think of using either `podman untag ...` to remove the name from an existing image or `podman build ...` to build the image from a Containerfile.  In both cases, it seems like the `podman images` JSON has a `Dangling: true` field, and in the latter you would have the base `fedora` image showing up in `podman images`.

Or did you use some other trick?  :)

debarshiray replied:
I took the liberty to add a `Dangling: true` field.

debarshiray commented on src/pkg/podman/imageImages_test.go:
```diff
@@ -0,0 +1,792 @@
+				"		\"RepoTags\": null," +
+				"		\"RepoDigests\": [" +
+				"			\"registry.fedoraproject.org/fedora-toolbox@sha256:0a51adc6bab55d49ff00da8aaad81ca1f02315511ed23d55ee5bbbe1a976a663\"," +
+				"			\"registry.fedoraproject.org/fedora-toolbox@sha256:8599f0c0d421c0dc01c4b7fb1c07b2780c0ab1931d0f22dd7b6da3b93ff6b77b\"" +
```
While playing with this, I learnt that the `RepoDigest` entries are duplicated if the second name is something like `localhost/fedora-toolbox:41`, but not if it's `registry.fedoraproject.org/fedora-toolbox:41-test` like in this test.  I didn't expect that.  :)

debarshiray commented on src/pkg/podman/imageImages_test.go:
```diff
@@ -0,0 +1,792 @@
+				"		\"Containers\": 1," +
+				"		\"Digest\": \"sha256:0a51adc6bab55d49ff00da8aaad81ca1f02315511ed23d55ee5bbbe1a976a663\"," +
+				"		\"History\": [" +
+				"			\"registry.fedoraproject.org/fedora-toolbox:41\"" +
```
Shouldn't the `History` field also have `registry.fedoraproject.org/fedora-toolbox:41-test`?

debarshiray replied:
I took the liberty to change this.

debarshiray commented on src/pkg/podman/imageImages_test.go:
```diff
@@ -0,0 +1,792 @@
+				assert.Equal(t, expect.isToolbx, image.IsToolbx())
+				assert.Equal(t, expect.labels, image.Labels())
+				assert.Equal(t, expect.names, image.Names())
+				assert.Equal(t, expect.repoTags, image.RepoTags())
```
I wonder if there's an easy way to test the `Created()` method.  It can be tricky because the relative time difference between the timestamp in the JSON and when the tests are run will keep increasing.  So, *5 minutes ago* will become *5 days ago*, etc.

I think we hit the same problem with tests for `containerInspect` and that's why it's not tested there either.  We can maybe simplify things a bit by using the same timestamp in the JSON for all the test cases.

It's not a big deal.  Just a thought.  I didn't investigate this much.  It's not a blocker for this PR.

debarshiray reviewed (commented):
I finally finished going through all the test cases for parsing the `podman images --format json` output.  The tests have revealed some mistakes elsewhere in the code, and I have one or two minor comments and questions about the changes here in this PR.

debarshiray commented on src/pkg/podman/imageImages_test.go:
```diff
@@ -0,0 +1,792 @@
+				"		]," +
+				"		\"Created\": 1759729685," +
+				"		\"CreatedAt\": \"2025-10-06T05:48:05Z\"" +
+				"	}" +
```
Finally, shouldn't there be two identical images in the `podman images` JSON when an image has two names?

debarshiray replied:
I took the liberty to change this.

debarshiray commented on src/pkg/podman/imageImages_test.go:
```diff
@@ -0,0 +1,792 @@
+			expects: []expect{
+				{
+					id:       "c49513deb6160607504d2c9abf9523e81c02f69ea479fd07572a7a32b50beab8",
+					isToolbx: false,
```
This made me realize that our claim of supporting Podman 1.6.4 is invalid.  :)

Due to commit [2369da5d31830e5c](https://github.com/containers/toolbox/commit/2369da5d31830e5cd3e1a13857a686365875ae61), Toolbx depends on the `podman images` JSON containing the labels, and this test case shows that they are not there with Podman 1.8.0.  :D

Anyway, it's not the fault of this PR. So there's no need to block on fixing this. It's a note to ourselves to do something about this dissonance. :)

debarshiray commented:
To keep things moving, I have split the changes introducing the `Image` interface and the `imageImages` concrete type, and the associated tests into https://github.com/containers/toolbox/pull/1779

debarshiray commented:
> To keep things moving, I have split the changes introducing the `Image` interface and the `imageImages` concrete type, and the associated tests into #1779

I took the liberty to rebase this PR on top of https://github.com/containers/toolbox/pull/1779 , which has now been merged into `main`.

debarshiray commented on src/pkg/podman/image.go:
```diff
@@ -37,10 +38,21 @@ type Images struct {
+type imageInspect struct {
+	created      string
+	entrypoint   []string
+	envVars      []string
```
It looks like `entrypoint` and `envVars` are not needed to parse the output of `podman inspect --type image` into an `Image` implementation.  I suppose you needed them for the changes in https://github.com/containers/toolbox/pull/1707 to address https://github.com/containers/toolbox/issues/1622 , right?  If so, it might be better to move these parts to the pull request that needs them.

More importantly, don't we need to add methods to the `Image` interface to cover these?

DaliborKr replied:
Yes, you are right. I prepared fields `entrypoint` and `envVars` in advance for #1707, so there is no use case for them right now. We can get rid of them now, and we can add them once we rewrite #1707. 

I am not sure about adding methods for them to the `Image` interface, since `imageImages`, which parses the `podman images` command introduced in #1779, does not provide `entrypoint` and `envVars` information.

DaliborKr replied:
I checked it out to be sure, and it is true. The `podman images --format json` does not provide properties for `entrypoint` and `envVars`.

debarshiray reviewed (commented):
Now onto the, hopefully, final round involving the `imageInspect` concrete type.  :)

debarshiray commented on src/pkg/podman/image.go:
```diff
@@ -37,10 +38,21 @@ type Images struct {
+	id           string
+	labels       map[string]string
+	namesHistory []string
+	repoTags     []string
```
While playing with the `podman inspect` JSON for images without names, it seems like the `RepoTags` field should be used for `Names()`, not `NamesHistory`.  eg., in such images `RepoTags` is always empty and `NamesHistory` always has some values.

Anyway, let's not block this pull request on this issue.  The existing code was ambiguous and it will be better to have this clarified before choosing to use `RepoTags` for `Names()`.
